src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java

```diff
@@ -75,7 +127,85 @@ private void submitRecordDeleteOperation() {}
 
   private void submitRecordReadOperation() {}
 
-  private void submitRecordCreateOperation() {}
+  boolean completeFailedOperation(final O op, final Throwable thrown) {
+    LogUtil.exception(Level.DEBUG, thrown, "{}: operation failed: {}", stepId, op);
+    return completeOperation(op, FAIL_UNKNOWN);
+  }
+
+  boolean completeOperation(final O op, final Operation.Status status) {
+    concurrencyThrottle.release();
+    op.status(status);
+    op.finishRequest();
+    op.startResponse();
+    op.finishResponse();
+    return handleCompleted(op);
+  }
+
+  private Properties createConfig(String nodeAddr) {
+    var producerConfig = new Properties();
+    producerConfig.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, nodeAddr);
+    producerConfig.put(ProducerConfig.BATCH_SIZE_CONFIG, this.batch);
+    producerConfig.put(ProducerConfig.MAX_REQUEST_SIZE_CONFIG, this.requestSizeLimit);
+    producerConfig.put(ProducerConfig.BUFFER_MEMORY_CONFIG, this.buffer);
+    producerConfig.put(ProducerConfig.COMPRESSION_TYPE_CONFIG, compression);
+    producerConfig.put(ProducerConfig.SEND_BUFFER_CONFIG, this.sndBuf);
+    producerConfig.put(ProducerConfig.LINGER_MS_CONFIG, this.linger);
+    producerConfig.put(ProducerConfig.RECEIVE_BUFFER_CONFIG, this.rcvBuf);
+    producerConfig.put(
+        ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,
+        "org.apache.kafka.common.serialization.StringSerializer");
+    producerConfig.put(
+        ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
+        "com.emc.mongoose.storage.driver.kafka.io.DataItemSerializer");
+    return producerConfig;
+  }
+
+  private void submitRecordCreateOperation(final DataOperation recordOp, String nodeAddr) {
+    try {
+      val config = configCache.computeIfAbsent(nodeAddr, this::createConfig);
+      val adminConfig =
+          adminClientCreateFuncCache.computeIfAbsent(config, AdminClientCreateFunctionImpl::new);
+      val adminClient = adminClientCache.computeIfAbsent(nodeAddr, adminConfig);
+      val producerConfig =
+          producerCreateFuncCache.computeIfAbsent(config, ProducerCreateFunctionImpl::new);
+      val kafkaProducer = producerCache.computeIfAbsent(nodeAddr, producerConfig);
+      val recordItem = recordOp.item();
+      val topicName = recordItem.name();
```

Comment:
The record item name will look like `/<somepath>/<somename>`. I suggest to use the `<somepath>` for the topic name and `<somename>` for the key. Example:
https://github.com/emc-mongoose/mongoose-storage-driver-pravega/blob/1f693952f82dfe86ff52f5bea7587d8816c46902/src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java#L467

Reply:
As I understand topicName = recordOp.dstPath(), key = recordItem,name()

Reply:
Yes, if dstPath doesn't include item name (I'm unsure)